Geometry_Engine/Query/RegionIntersection.cs

```diff
@@ -0,0 +1,111 @@
+/*
+ * This file is part of the Buildings and Habitats object Model (BHoM)
+ * Copyright (c) 2015 - 2023, the respective contributors. All rights reserved.
+ *
+ * Each contributor holds copyright over their respective contributions.
+ * The project versioning (Git) records all such contribution source information.
+ *                                           
+ *                                                                              
+ * The BHoM is free software: you can redistribute it and/or modify         
+ * it under the terms of the GNU Lesser General Public License as published by  
+ * the Free Software Foundation, either version 3.0 of the License, or          
+ * (at your option) any later version.                                          
+ *                                                                              
+ * The BHoM is distributed in the hope that it will be useful,              
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of               
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the                 
+ * GNU Lesser General Public License for more details.                          
+ *                                                                            
+ * You should have received a copy of the GNU Lesser General Public License     
+ * along with this code. If not, see <https://www.gnu.org/licenses/lgpl-3.0.html>.      
+ */
+
+using BH.Engine.Base;
+using BH.oM.Geometry;
+using BH.oM.Base.Attributes;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.ComponentModel;
+using BH.oM.Quantities.Attributes;
+using BH.oM.Geometry.CoordinateSystem;
+
+namespace BH.Engine.Geometry
+{
+    public static partial class Query
+    {
+        /***************************************************/
+        /**** Public Methods                            ****/
+        /***************************************************/
+
+        [Description("Find segments of a line lying inside the region defined by a polyline.")]
+        [Input("line", "A line to check for intersections with a region defined by the input polylines.")]
+        [Input("pLine", "A polyline defining a closed region that the input line potentially intersects.")]
+        [Input("tolerance", "Minimum length required of new intersection lines.")]
+        [Output("intersections", "Segments of the input line that lie inside the region defined by the input polyline.")]
+        public static List<Line> RegionIntersection(this Polyline region, Line line, double minLineLength = Tolerance.Distance, bool acceptOnEdge = true)
```

Comment:
I compared my method and yours and couldn't see a big potential performance difference though. My method has slightly more overhead because it needs to do 2 extra things:
- Only return lines longer than the minLineLength input
- Capture lines formed by an intersection point and 1 of the 2 endpoints of the cutting line if these endpoints are inside the polyline region.

If this is too specific to the problem it solves, I can move the method to the tool's repo.

![image](https://github.com/BHoM/BHoM_Engine/assets/102604891/37c98880-af93-4e3e-bcc9-9bf1cc76e545)

Reply:
I slightly tweaked the method to do exactly same thing as yours:
```C#
public static List<Line> RegionIntersection2(this ICurve curve, Line line, bool acceptOnEdges = false, double minLineLength = Tolerance.Distance, double tolerance = Tolerance.Distance)
        {
            List<Point> intPts = Query.ILineIntersections(curve, line, true, tolerance);
            intPts.Add(line.Start);
            intPts.Add(line.End);
            intPts = intPts.CullDuplicates(tolerance);
            intPts = intPts.SortCollinear(tolerance);

            List<Line> result = new List<Line>();
            for (int i = 0; i < intPts.Count - 1; i++)
            {
                if (intPts[i].Distance(intPts[i + 1]) >= minLineLength && curve.IIsContaining(new List<Point> { intPts.Skip(i).Take(2).Average() }, acceptOnEdges, tolerance))
                    result.Add(new Line { Start = intPts[i], End = intPts[i + 1] });
            }

            return result;
        }
```

We have some 30% performance gain (see below) + the code is cleaner imo.
![image](https://github.com/BHoM/BHoM_Engine/assets/26874773/2fe196ef-84f0-4788-9d5c-b93d89679e6c)